umapi_client/functional.py

```diff
@@ -308,6 +309,61 @@ def delete_account(self):
         self.append(removeFromDomain={})
         return None
 
+    def maybe_split_groups(self, max_groups):
+        """
+        Check if group lists in add/remove directives should be split and split them if needed
+        :param max_groups: Max group list size
+        :return: True if at least one command was split, False if none were split
+        """
+        new_commands = []
+        # return True if we split at least once
+        maybe_split = False
+        for command in self.commands:
+            do_split = False
+            for verb, verb_commands in six.iteritems(command):
+                for group_type, groups in six.iteritems(verb_commands):
+                    if len(groups) > max_groups:
+                        do_split = True
+            if do_split:
+                maybe_split = True
+                new_commands += self._split_groups(command, max_groups)
+            else:
+                new_commands.append(command)
+        self.commands = new_commands
+        return maybe_split
+
+    def _split_groups(self, command, max_groups):
+        """
+        Split long group lists of individual command
+        Creates additional commands for each split and recursively splits new groups until no commands exceed group
+        limit.  Assumes that at least one split should occur.
+        :param command: Single command containing one or more group lists to add, remove, etc
+        :param max_groups: Max group list size
+        :return: List of at least two commands that resulted from split
+        """
+        new_command = copy.deepcopy(command)
```

Comment:
yikes, no.  Again, these are dictionaries, and you don't need to do a deep copy, because there is no shared structure.  Let's assume you are passing in the verb of the command:
```python
new_commands = [command]
while True:
    new_command = {verb: {}}:
    for group_type, groups in six.iteritems(command[verb]):
        if len(groups) > max_groups:
            command[verb][group_type], new_command[verb][group_type] = groups[0:max_groups], groups[max_groups:]
    if new_command[verb]:
        new_commands.append(new_command)
    else:
        return new_commands
```
This works even if there are no group lists that are longer than max_groups, which allows you to actually combine these two functions by setting maybe_split to true in the `if` part.  That's what I really meant by combining the test and the split.  And it's a lot shorter than what you have.

Reply:
I chose to deep copy because the command dictionary can directly or indirectly contain dictionaries and lists.  A shallow copy won't duplicate those objects; just references.  I can construct a new command dict as you're doing here, but my original intent was to handle complex, multi-key commands that are allowed by the API.  

If we decide to not support multi-key commands, I'll implement the simpler method in your example.

Reply:
I simplified the split logic based on the assumption that there will be one add, remove, etc key per step.  The update you suggested doesn't work when group lists need additional splitting - i.e. when list size is > 2*max_group_size.  I had to tweak it a bit.